packages/composable-controller/src/ComposableController.ts

```diff
@@ -168,33 +181,34 @@ export class ComposableController extends BaseController<
   /**
    * Constructor helper that subscribes to child controller state changes.
    * @param controller - Controller instance to update
-   * TODO: Remove `isBaseControllerV1` branch once `BaseControllerV2` migrations are completed for all controllers.
    */
   #updateChildController(controller: ControllerInstance): void {
+    if (!isBaseController(controller) && !isBaseControllerV1(controller)) {
+      throw new Error(
+        'Invalid controller: controller must extend from BaseController or BaseControllerV1',
+      );
+    }
+
     const { name } = controller;
     if (isBaseControllerV1(controller)) {
       controller.subscribe((childState) => {
-        this.update((state) => ({
-          ...state,
-          [name]: childState,
-        }));
+        this.update((state) => {
+          Object.assign(state, { [name]: childState });
+        });
       });
-    } else if (isBaseController(controller)) {
+    }
+    if (
+      (isBaseControllerV1(controller) && 'messagingSystem' in controller) ||
+      isBaseController(controller)
+    ) {
       this.messagingSystem.subscribe(
```

Comment:
If a BaseControllerV1 controller has a `messagingSystem` property, could the state of this controller could be updated twice per update? If so, do we still need an `else if`?

Reply:
Great point! Fixed here: https://github.com/MetaMask/core/pull/3964/commits/811a94a17ecf76c290a621be19becd0696ed7d55